modules/launchd/default.nix

```diff
@@ -0,0 +1,192 @@
+  };
+
+  toAgent = config: pkgs.writeText "${config.Label}.plist" (toPlist { } config);
+
+  agentPlists =
+    mapAttrs' (n: v: nameValuePair "${v.config.Label}.plist" (toAgent v.config))
+    (filterAttrs (n: v: v.enable) cfg.agents);
+
+  agentsDrv = pkgs.runCommand "home-manager-agents" {
+    srcs = attrValues agentPlists;
+    dsts = attrNames agentPlists;
+  } ''
+    mkdir -p "$out"
+
+    if [[ -n "$srcs" ]]; then
+      for (( i=0; i < "''${#srcs[@]}"; i+=1 )); do
+        src="''${srcs[i]}"
+        dst="''${dsts[i]}"
+        ln -s "$src" "$out/$dst"
+      done
+    fi
+  '';
+in {
+  meta.maintainers = with maintainers; [ midchildan ];
+
+  options.launchd = {
+    enable = mkOption {
+      type = types.bool;
+      default = isDarwin;
+      defaultText = literalExpression "pkgs.stdenv.hostPlatform.isDarwin";
+      description = ''
+        Whether to enable Home Manager to define per-user daemons by making use
+        of launchd's LaunchAgents.
+      '';
+    };
+
+    agents = mkOption {
+      type = with types; attrsOf (submodule launchdConfig);
+      default = { };
+      description = "Define LaunchAgents.";
+    };
+  };
+
+  config = mkMerge [
+    {
+      assertions = [{
+        assertion = (cfg.enable && agentPlists != { }) -> isDarwin;
+        message = let names = lib.concatStringsSep ", " (attrNames agentPlists);
+        in "Must use Darwin for modules that require Launchd: " + names;
+      }];
+    }
+
+    (mkIf isDarwin {
+      home.extraBuilderCommands = ''
+        ln -s "${agentsDrv}" $out/LaunchAgents
+      '';
+
+      home.activation.checkLaunchAgents =
+        hm.dag.entryBefore [ "writeBoundary" ] ''
+          checkLaunchAgents() {
+            local oldDir="$(readlink -m "$oldGenPath/LaunchAgents")"
+            local newDir=${escapeShellArg agentsDrv}
+            local dstDir=${escapeShellArg dstDir}
+
+            local oldSrcPath newSrcPath dstPath agentFile agentName
+
+            find -L "$newDir" -maxdepth 1 -name '*.plist' -type f -print0 \
+                | while IFS= read -rd "" newSrcPath; do
+              agentFile="''${newSrcPath##*/}"
+              agentName="''${agentFile%.plist}"
+              dstPath="$dstDir/$agentFile"
+              oldSrcPath="$oldDir/$agentFile"
+
+              if [[ ! -e "$dstPath" ]]; then
+                continue
+              fi
+
+              if ! cmp --quiet "$oldSrcPath" "$dstPath"; then
+                errorEcho "Existing file '$dstPath' is in the way of '$newSrcPath'"
+                exit 1
+              fi
+            done
+          }
+
+          checkLaunchAgents
+        '';
+
+      # NOTE: Launch Agent configurations can't be symlinked from the Nix store
+      # because it needs to be owned by the user running it.
+      home.activation.setupLaunchAgents =
+        hm.dag.entryAfter [ "writeBoundary" ] ''
+          setupLaunchAgents() {
+            local oldDir="$(readlink -m "$oldGenPath/LaunchAgents")"
```

Comment:
I'm wondering if this may cause problems on the first activation, when `$oldGenPath` is unset?

Btw, the general recommendation is to declare the variable and assign it in separate statements. I.e. something like

``` sh
local oldDir
oldDir="$(readline …)"
```

See [SC2155](https://github.com/koalaman/shellcheck/wiki/SC2155) for rationale.

Reply:
I added more conditional statements to deal with an empty `$oldGenPath` and also separated variable declarations and assignments.